bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataGrpcClient.java

```diff
@@ -267,18 +242,40 @@ public void run() {
       }
     }, executorService);
 
+    int timeout = retryOptions.getReadPartialRowTimeoutMillis();
+
+    int streamingBufferSize = retryOptions.getStreamingBufferSize();
+    ResponseQueueReader responseQueueReader =
+        new ResponseQueueReader(timeout, streamingBufferSize, streamingBufferSize, readRowsCall);
+
     StreamingBigtableResultScanner resultScanner =
-        new StreamingBigtableResultScanner(
-          reservedChannel,
-          retryOptions.getStreamingBufferSize(),
-          retryOptions.getReadPartialRowTimeoutMillis(),
-          cancellationToken);
+        new StreamingBigtableResultScanner(reservedChannel, responseQueueReader, cancellationToken);
 
     clientCallService.asyncServerStreamingCall(
         readRowsCall,
         request,
-        new ReadRowsStreamObserver(resultScanner));
+        streamingBufferSize,
+        createClientCallListener(resultScanner));
 
     return resultScanner;
   }
+
+  private ClientCall.Listener<ReadRowsResponse> createClientCallListener(
+      final StreamingBigtableResultScanner resultScanner) {
+    return new ClientCall.Listener<ReadRowsResponse>() {
+      @Override
+      public void onMessage(ReadRowsResponse readRowResponse) {
+        resultScanner.addResult(readRowResponse);
```

Comment:
I think we need to request x when:
(capacity - (queue size + outstanding requests) < x) && !(finished)

As you said, the most likely place for that formula to change is when the user thread is hungry for more.  I would think that the most likely place to determine if a "user thread is hungry for more" is on the user thread, not the producer/onMessage thread.

Consider this scenario.  What happens if onMessage is called enough times to fill the queue before the user thread can pull things off?